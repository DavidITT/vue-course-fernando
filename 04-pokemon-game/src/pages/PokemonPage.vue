<template>
  <div class="d-flex flex-column justify-content-center align-items-center vh-100">

    <div v-if="pokemon">
      <h1 class="mb-5">Â¿Quien es este pokemon?</h1>
      <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon"/>
      <PokemonOptions :options="pokemonArr" @selection="checkAnswer"/>
      <button v-if="answer" class="btn btn-primary" @click="resetGame">Reiniciar</button>
    </div>

    <div v-else class="d-flex flex-column">
      <div class="spinner-border mb-3 text-info" style="width: 6rem; height: 6rem;" role="status">
        <span class="visually-hidden">Loading...</span>
      </div>
      <span>Espere ...</span>
    </div>

    <div v-if="answer" class="alert d-flex align-items-center w-100 fixed-top" :class="classMessage" role="alert">
      <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="bi bi-check-square-fill"
           viewBox="0 0 16 16">
        <path
            d="M2 0a2 2 0 0 0-2 2v12a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V2a2 2 0 0 0-2-2H2zm10.03 4.97a.75.75 0 0 1 .011 1.05l-3.992 4.99a.75.75 0 0 1-1.08.02L4.324 8.384a.75.75 0 1 1 1.06-1.06l2.094 2.093 3.473-4.425a.75.75 0 0 1 1.08-.022z"/>
      </svg>
      <div class="mx-3">
        {{ message }}
      </div>
    </div>
  </div>
</template>

<script>
import PokemonPicture from '@/components/PokemonPicture'
import PokemonOptions from '@/components/PokemonOptions'
import getPokemonsOptions from '@/helpers/getPokemonOptions'

export default {
  name: "PokemonPage",

  data() {
    return {
      pokemonArr: [],
      pokemon: null,
      showPokemon: false,
      answer: false,
      message: ' ',
      classMessage: ''
    }
  },
  methods: {
    async mixPokemonArray() {
      this.pokemonArr = await getPokemonsOptions()
      const randomInt = Math.floor(Math.random() * 4)
      this.pokemon = this.pokemonArr[randomInt]
    },
    checkAnswer(pokemonId) {
      if (this.pokemon.id === pokemonId) {
        this.showPokemon = true
        this.answer = true
        this.message = 'Respuesta correcta'
        this.classMessage = 'alert-success'
      } else {
        this.answer = true
        this.message = 'Respuesta incorrecta'
        this.classMessage = 'alert-danger'
      }
    },
    resetGame() {
      this.pokemonArr = []
      this.pokemon = null
      this.showPokemon = false
      this.answer = false
      this.message = ' '
      this.classMessage = ''
      this.mixPokemonArray()

    }
  },
  components: {
    PokemonPicture,
    PokemonOptions
  },
  mounted() {
    this.mixPokemonArray()
  }
}
</script>

