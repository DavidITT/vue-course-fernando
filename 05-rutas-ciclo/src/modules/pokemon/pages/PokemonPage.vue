<template>
  <h1>Pokemon: #{{ id }}</h1>
  <div v-if="pokemon">
    <img :src="pokemon.sprites.front_default" :alt="pokemon.name" width="200" height="200">
  </div>
  <!--  <h1>Pokemon: #{{ $route.params.id }}</h1>-->
</template>

<script>
export default {
  name: "PokemonPage",

  props: {
    id: {
      type: Number,
      required: true
    },
  },

  data() {
    return {
      pokemon: null
    }
  },

  methods: {
    async getPokemon() {
      try {
        this.pokemon = await fetch(`https://pokeapi.co/api/v2/pokemon/${this.id}`).then(r => r.json())
      } catch (error) {
        this.$router.push('/')
        console.log('No hay nada que hacer aqui')
      }
    }
  },
  watch: {
    id() {
      this.getPokemon()
    }
  },
  created() {
    //const {id} = this.$route.params.id
    //console.log(id)
    //this.id = id
    this.getPokemon()
  }
}
</script>

<style scoped>

</style>