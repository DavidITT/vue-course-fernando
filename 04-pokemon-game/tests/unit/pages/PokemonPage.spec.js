import {shallowMount, mount} from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage'
import {pokemons} from "../mocks/pokemons.mock";

describe('PokemonPage Component', () => {
    let wrapper


    beforeEach(() => {
        wrapper = shallowMount(PokemonPage)
    })

    test('Debe de hacer match con el snapshoot', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })

    test('Debe de llamar el mixPokemonArray al montar', () => {
        const mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray')
        shallowMount(PokemonPage)

        expect(mixPokemonArraySpy).toHaveBeenCalled()

    })

    test('Debe de hacer match con el snapshot cuadno cargan los pokemons', () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    answer: false,
                    message: ' ',
                    classMessage: ''
                }
            }
        })
        expect(wrapper.html()).toMatchSnapshot()
    })

    test('Debe de mostrar los componentes de PokemonPicture y PokemonOptions', () => {

        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    answer: false,
                    message: ' ',
                    classMessage: ''
                }
            }
        })

        const picture = wrapper.find('pokemon-picture-stub')
        const options = wrapper.find('pokemon-options-stub')

        expect(picture.exists()).toBeTruthy()
        expect(options.exists()).toBeTruthy()


        expect(picture.attributes('pokemonid')).toBe('1')
        expect(options.attributes('options')).toBeTruthy()

    })

    test('Pruebas con checkAnswer', async () => {

        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    answer: false,
                    message: ' ',
                    classMessage: ''
                }
            }
        })

       await wrapper.vm.checkAnswer(1)

        expect(wrapper.find('#answer').exists()).toBeTruthy()
        expect(wrapper.vm.showPokemon).toBe(true)
        expect(wrapper.find('#answer').text()).toBe('Respuesta correcta')

        await wrapper.vm.checkAnswer(2)

        expect(wrapper.vm.message).toBe('Respuesta incorrecta')


    })
})