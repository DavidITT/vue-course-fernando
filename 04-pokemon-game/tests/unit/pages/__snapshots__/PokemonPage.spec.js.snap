// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component Debe de hacer match con el snapshoot 1`] = `
<div class="d-flex flex-column justify-content-center align-items-center vh-100">
  <div class="d-flex flex-column">
    <div class="spinner-border mb-3 text-info" style="width: 6rem; height: 6rem;" role="status"><span class="visually-hidden">Loading...</span></div><span>Espere ...</span>
  </div>
  <!--v-if-->
</div>
`;

exports[`PokemonPage Component Debe de hacer match con el snapshot cuadno cargan los pokemons 1`] = `
<div class="d-flex flex-column justify-content-center align-items-center vh-100">
  <div>
    <h1 class="mb-5">Â¿Quien es este pokemon?</h1>
    <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
    <pokemon-options-stub options="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
    <!--v-if-->
  </div>
  <!--v-if-->
</div>
`;
